### PR TITLE
Scheme prenormalization

### PR DESCRIPTION
**Pull request by leogrim**
@ymatsuda @yingjieMiao @eishay

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/normalizer/URINormalizer.scala [hunk 1]
```diff
@@ -6,28 +6,16 @@ trait URINormalizer extends PartialFunction[URI, URI]
...
-          val delta = path.length - path.toLowerCase.length // in case the case conversion changed the length
-          path.substring(0, m.start + delta) + "/"
-        case _ => path
-      }
-    }
-  }
+  private def applyAll(uri: URI, normalizers: Seq[StaticNormalizer]) = normalizers.foldLeft(uri)((u, n) => n.applyOrElse(u, identity[URI]))
+  private def applyFirst(uri: URI, normalizers: Seq[StaticNormalizer]) = normalizers.find(_.isDefinedAt(uri)).map(_.apply(uri)).get
+
 }
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/normalizer/URINormalizer.scala:21]: @ymatsuda refactored some code here, hope it makes sense

File: kifi-backend/modules/shoebox/app/com/keepit/normalizer/NormalizationService.scala [hunk 2]
```diff
@@ -109,10 +117,10 @@ class NormalizationServiceImpl @Inject() (
           if (currentReference.normalization == Some(strongerCandidate.normalization)) assert(currentReference.url != strongerCandidate.url)
 
           db.readOnly { implicit session =>
-            priorKnowledge(strongerCandidate) match {
-              case PriorKnowledge.ACCEPT => Future.successful((Some(strongerCandidate), weakerCandidates))
-              case PriorKnowledge.REJECT => findCandidate(weakerCandidates)
-              case PriorKnowledge.Check(contentCheck) =>
+            oracle(strongerCandidate) match {
```
Comment (gmethvin) [inline kifi-backend/modules/shoebox/app/com/keepit/normalizer/NormalizationService.scala:125]: nice naming

File: kifi-backend/modules/shoebox/app/com/keepit/normalizer/Action.scala [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package com.keepit.normalizer
+
+import com.keepit.model.NormalizedURI
+
+sealed trait Action
+case object ACCEPT extends Action
```
Comment (gmethvin) [inline kifi-backend/modules/shoebox/app/com/keepit/normalizer/Action.scala]: I prefer `Accept` and `Reject` for class names over all caps.

File: kifi-backend/modules/shoebox/app/com/keepit/normalizer/NormalizationService.scala [hunk 4]
```diff
@@ -29,10 +30,17 @@ class NormalizationServiceImpl @Inject() (
...
   uriIntegrityPlugin: UriIntegrityPlugin,
+  priorKnowledge: PriorKnowledge,
   scraperPlugin: ScraperPlugin,
   healthcheckPlugin: HealthcheckPlugin) extends NormalizationService with Logging {
 
-  def normalize(uriString: String)(implicit session: RSession): String = normalizedURIRepo.getByUri(uriString).map(_.url).getOrElse(Prenormalizer(uriString))
+  def normalize(uriString: String)(implicit session: RSession): String = normalizedURIRepo.getByUri(uriString).map(_.url).getOrElse(prenormalize(uriString))
+  def prenormalize(uriString: String)(implicit session: RSession): String = {
+    val withStandardPrenormalizationOption = URI.safelyParse(uriString).map(Prenormalizer)
+    val withPreferredSchemeOption = for { prenormalized <- withStandardPrenormalizationOption; schemeNormalizer <- priorKnowledge.getPreferredSchemeNormalizer(uriString) } yield schemeNormalizer(prenormalized)
```
Comment (gmethvin) [inline kifi-backend/modules/shoebox/app/com/keepit/normalizer/NormalizationService.scala]: it's a bit more readable to write like

``` scala
for {
  prenormalized <- URI.safelyParse(uriString).map(Prenormalizer)
  schemeNormalizer <- priorKnowledge.getPreferredSchemeNormalizer(uriString)
} yield schemeNormalizer(prenormalized)
```

File: kifi-backend/modules/shoebox/app/com/keepit/normalizer/Action.scala (hunk 3 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/normalizer/Action.scala]: Even for constant objects?

**Comment by yingjieMiao** [comment]
lgtm

File: kifi-backend/modules/shoebox/app/com/keepit/normalizer/Action.scala (hunk 3 above)
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/normalizer/Action.scala]: I guess that's the convention here. See those message objects sent to actors.

**Comment by leogrim** [comment]
@ymatsuda @yingjieMiao @2is10 
Ok to merge this? 
This will lead to the end of a couple of repo/tables (Unscrapable, UrlPattern), duplicated admin interfaces and similar controllers. Note that the extension still relies on UrlPattern to get know when to show the slider, but as soon as I have entered the rules from that table in the new UrlPatternRule repo, I'll do a general code + db cleanup.

Also completes the original intent of enabling scheme normalization for selected sites that led to nice improvements to the normalization code =)

**Comment by yingjieMiao** [comment]
I am good with that.

**Comment by 2is10** [comment]
I didn’t realize there would need to be columns that make sense for some rows and not others.

My vote would be to have separate tables for separate uses. The fact that two tables both contain regular expressions that match URLs doesn’t necessarily mean that they should be combined into one table.

If you feel that one or more tables are named poorly (e.g. `url_patterns` is not specific enough), I’d be fine with renaming.

**Comment by leogrim** [comment]
All the current tables usually contain the same problematic sites (Google's, Facebook, LinkedIn...). I'm not sure what you mean regarding the columns not making sense for all rows, the bonus here is that we get a clear view of what's happening per website, which seems easier for debugging user feedback, doesn't it?

**Comment by 2is10** [comment]
I only know what’s in one of the current tables. I incorrectly surmised
from our discussion that we’d keep the same user interface for
viewing/editing patterns for auto-showing the keeper and have a separate
admin UI for search-related patterns, both of which happened to be backed
by the same table. I didn’t realize you intended to reuse the same rows
(patterns) for multiple purposes and build a single, more complex user
interface. This will be better for people who understand the whole system,
but might be slower/harder for people trying to make a simple change to one
aspect, like auto-showing the keeper. I don’t feel strongly either way, so
sure, go ahead.

**Comment by leogrim** [comment]
Ok! Well the UI is quite straightforward (yours + a few explicit columns with default values provided for the creation of a new row). I'll push this and if you really find it too complex, I can break it up. The trick is that using the same repo, it's nice to be able to see all the regexes at a glance to make sure there's no conflict (I still have assertions that check for that at runtime).

**Pull request merged**